src/conv.jl

```diff
@@ -329,6 +329,62 @@ for (front_name, backend) in (
     end
 end
 
+# direct function forwarding definition
+function ∇conv_data!(out::AbstractArray{yT,N}, in1::AbstractArray{T1,N},
+                     in2::AbstractArray{T2,N}, cdims::C; kwargs...) where {yT, T1, T2, N, C <: ConvDims}
+    if yT == Float64  # warn for Float32 + accidental Float64, but don't print warning for ForwardDiff.Dual
+        @warn string("Slow fallback implementation invoked for ", string(front_name), "!  ",
+                "You probably don't want this; check your datatypes.") yT T1 T2 maxlog=1
+    end
+
+    dx_cs = Iterators.partition(1:size(out, 4),
+                channels_in(cdims) ÷ groupcount(cdims))
+    w_cs = Iterators.partition(1:size(in2, 5),
+                channels_out(cdims) ÷ groupcount(cdims))
+    dy_cs = Iterators.partition(1:size(in1, 4),
+                channels_out(cdims) ÷ groupcount(cdims))
+    cdims2 = basetype(C)(cdims,
+            G = 1,
+            C_in = channels_in(cdims) ÷ groupcount(cdims),
+            C_out = channels_out(cdims) ÷ groupcount(cdims))
+
+    Threads.@sync for (xc, yc, wc) in zip(dx_cs, dy_cs, w_cs)
+        dxv = @view out[ntuple(i -> i == 4 ? xc : Colon(), 5)...]
+        dyv = @view in1[ntuple(i -> i == 4 ? yc : Colon(), 5)...]
+        wv = @view in2[ntuple(i -> i == 5  ? wc : Colon(), 5)...]
+        Threads.@spawn ∇conv_data_direct!(dxv, dyv, wv, cdims2; kwargs...)
+    end
+
+    return out
+end
+
+function ∇conv_filter!(out::AbstractArray{yT,N}, in1::AbstractArray{T1,N},
+                       in2::AbstractArray{T2,N}, cdims::C; kwargs...) where {yT, T1, T2, N, C <: ConvDims}
+    if yT == Float64  # warn for Float32 + accidental Float64, but don't print warning for ForwardDiff.Dual
+        @warn string("Slow fallback implementation invoked for ", string(front_name), "!  ",
+                "You probably don't want this; check your datatypes.") yT T1 T2 maxlog=1
+    end
+    dw_cs = Iterators.partition(1:size(out, 5),
+                channels_out(cdims) ÷ groupcount(cdims))
+    dy_cs = Iterators.partition(1:size(in2, 4),
+                channels_out(cdims) ÷ groupcount(cdims))
+    x_cs = Iterators.partition(1:size(in1, 4),
+                channels_in(cdims) ÷ groupcount(cdims))
+    cdims2 = basetype(C)(cdims,
+            G = 1,
+            C_in = channels_in(cdims) ÷ groupcount(cdims),
+            C_out = channels_out(cdims) ÷ groupcount(cdims))
+
+    Threads.@sync for (wc, xc, yc) in zip(dw_cs, x_cs, dy_cs)
+        x = @view in1[ntuple(i -> i == 4 ? xc : Colon(), 5)...]
+        dy = @view in2[ntuple(i -> i == 4 ? yc : Colon(), 5)...]
+        dw = @view out[ntuple(i -> i == 5 ? yc : Colon(), 5)...]
+        Threads.@spawn ∇conv_filter_direct!(dw, x, dy, cdims2; kwargs...)
+    end
+
+    return out
+end
```

Comment:
This duplication with the code above is a little unfortunate. Is there any way we could reuse some of it? Maybe loop over `(:direct, :im2col)`?

Reply:
Yes, I thought the same. I was trying to implement something like this, but was having some problems with the function signatures, since for the im2col all types must be the same and from the list of possible types. I'm not used to Julia's metaprogramming so I'm learning a little to try come up with a solution.

Reply:
Yes, I was trying to implement this, but I'm not used to Julia's metaprogramming, so I'm doing some catch up to try coming up with a solution.

I was having some trouble with the function signatures, since for im2col all types must be the same and part of G.

I tried initially have a third element in the for loop tuple with the signature types, but then it gets messy when doing the where statement.